File: src/server/codex/images.ts
import "server-only";

import { HttpError } from "@/src/server/http/errors";

const DEFAULT_IMAGES_MAIN_MODEL = "gpt-5.4-mini";
const DEFAULT_IMAGES_TOOL_MODEL = "gpt-image-2";
const MAX_IMAGE_EDIT_FILES = 10;
const MAX_IMAGE_EDIT_FILE_BYTES = 20 * 1024 * 1024;

type ImageAction = "generate" | "edit";
type ImageStreamPrefix = "image_generation" | "image_edit";

export type CodexImagesRequest = {
  payload: Record<string, unknown>;
  responseFormat: "b64_json" | "url";
  stream: boolean;
  model: string;
  requestBody: unknown;
  streamPrefix: ImageStreamPrefix;
};

type ImageCallResult = {
  result: string;
  revisedPrompt: string;
  outputFormat: string;
  size: string;
  background: string;
  quality: string;
};

type OutputItemRepairState = {
  byIndex: Map<number, Record<string, unknown>>;
  fallback: Record<string, unknown>[];
};

export function buildImagesGenerationsRequest(
  input: Record<string, unknown>,
): CodexImagesRequest {
  const imageModel = normalizeImagesModel(input.model);
  const prompt = stringValue(input.prompt);
  if (!prompt) {
    throw new HttpError(
      400,
      "missing_image_prompt",
      "Invalid request: prompt is required",
    );
  }

  const responseFormat = normalizeImagesResponseFormat(input.response_format);
  const tool = imageGenerationTool("generate", imageModel, input);
  const payload = buildImagesResponsesPayload(prompt, [], tool);

  return {
    payload,
    responseFormat,
    stream: Boolean(input.stream),
    model: stringValue(payload.model) || DEFAULT_IMAGES_MAIN_MODEL,
    requestBody: input,
    streamPrefix: "image_generation",
  };
}

export function buildImagesEditsJsonRequest(
  input: Record<string, unknown>,
): CodexImagesRequest {
  const imageModel = normalizeImagesModel(input.model);
  const prompt = stringValue(input.prompt);
  if (!prompt) {
    throw new HttpError(
      400,
      "missing_image_prompt",
      "Invalid request: prompt is required",
    );
  }

  const images = collectJsonImages(input);
  if (images.length === 0) {
    throw new HttpError(
      400,
      "missing_image_input",
      "Invalid request: images[].image_url is required (file_id is not supported)",
    );
  }

  const mask = imageUrlFromRecord(input.mask);
  const responseFormat = normalizeImagesResponseFormat(input.response_format);
  const tool = imageGenerationTool("edit", imageModel, input);
  if (mask) {
    tool.input_image_mask = { image_url: mask };
  } else if (isRecord(input.mask) && stringValue(input.mask.file_id)) {
    throw new HttpError(
      400,
      "unsupported_image_mask_file",
      "Invalid request: mask.file_id is not supported (use mask.image_url instead)",
    );
  }

  const payload = buildImagesResponsesPayload(prompt, images, tool);
  return {
    payload,
    responseFormat,
    stream: Boolean(input.stream),
    model: stringValue(payload.model) || DEFAULT_IMAGES_MAIN_MODEL,
    requestBody: input,
    streamPrefix: "image_edit",
  };
}

export async function buildImagesEditsMultipartRequest(
  formData: FormData,
): Promise<CodexImagesRequest> {
  const imageModel = normalizeImagesModel(formData.get("model"));
  const prompt = stringValue(formData.get("prompt"));
  if (!prompt) {
    throw new HttpError(
      400,
      "missing_image_prompt",
      "Invalid request: prompt is required",
    );
  }

  const bracketFiles = formData.getAll("image[]").filter(isFileLike);
  const files =
    bracketFiles.length > 0
      ? bracketFiles
      : formData.getAll("image").filter(isFileLike);
  if (files.length === 0) {
    throw new HttpError(
      400,
      "missing_image_input",
      "Invalid request: image is required",
    );
  }
  if (files.length > MAX_IMAGE_EDIT_FILES) {
    throw new HttpError(
      413,
      "too_many_image_inputs",
      `At most ${MAX_IMAGE_EDIT_FILES} image files are allowed`,
    );
  }

  const images = await Promise.all(files.map(fileToDataUrl));
  const responseFormat = normalizeImagesResponseFormat(
    formData.get("response_format"),
  );
  const tool = imageGenerationTool(
    "edit",
    imageModel,
    formRecordFromData(formData),
  );
  const mask = formData.get("mask");
  if (isFileLike(mask)) {
    tool.input_image_mask = { image_url: await fileToDataUrl(mask) };
  }

  const payload = buildImagesResponsesPayload(prompt, images, tool);
  return {
    payload,
    responseFormat,
    stream: booleanFormValue(formData.get("stream")),
    model: stringValue(payload.model) || DEFAULT_IMAGES_MAIN_MODEL,
    requestBody: {
      model: imageModel,
      prompt,
      image_count: images.length,
      has_mask: isFileLike(mask),
      response_format: responseFormat,
      stream: booleanFormValue(formData.get("stream")),
    },
    streamPrefix: "image_edit",
  };
}

export function buildImagesApiResponseFromSseText(
  text: string,
  responseFormat: "b64_json" | "url",
) {
  const outputItems = createOutputItemRepairState();
  for (const frame of parseSseFrames(text)) {
    const event = parseSseJson(frame);
    if (!event) {
      continue;
    }
    if (collectOutputItemDone(event, outputItems)) {
      continue;
    }
    if (event.type !== "response.completed") {
      continue;
    }
    const completed = extractImagesFromCompletedEvent(event, outputItems);
    return buildImagesApiResponse(completed, responseFormat);
  }
  throw new HttpError(
    502,
    "image_generation_incomplete",
    "Upstream stream ended before image generation completed",
  );
}

export function createImagesSseStream(
  upstreamBody: ReadableStream<Uint8Array>,
  input: {
    responseFormat: "b64_json" | "url";
    streamPrefix: ImageStreamPrefix;
    onCompleted: () => void;
    onError: (error: unknown) => void;
    onFirstEvent?: () => void;
  },
) {
  const decoder = new TextDecoder();
  const encoder = new TextEncoder();
  let buffer = "";
  let completed = false;
  let sawEvent = false;
  const outputItems = createOutputItemRepairState();

  function emit(
    controller: TransformStreamDefaultController<Uint8Array>,
    eventName: string,
    data: unknown,
  ) {
    controller.enqueue(
      encoder.encode(`event: ${eventName}\ndata: ${JSON.stringify(data)}\n\n`),
    );
  }

  function emitError(
    controller: TransformStreamDefaultController<Uint8Array>,
    error: unknown,
  ) {
    const message = error instanceof Error ? error.message : String(error);
    emit(controller, "error", {
      error: {
        message,
        type: "stream_error",
        code: "upstream_image_stream_error",
      },
    });
  }

  function fail(
    controller: TransformStreamDefaultController<Uint8Array>,
    error: unknown,
  ) {
    if (completed) {
      return;
    }
    completed = true;
    input.onError(error);
    emitError(controller, error);
  }

  function processFrame(
    frame: string,
    controller: TransformStreamDefaultController<Uint8Array>,
  ) {
    if (completed) {
      return;
    }
    const event = parseSseJson(frame);
    if (!event) {
      return;
    }
    if (!sawEvent) {
      sawEvent = true;
      input.onFirstEvent?.();
    }

    if (collectOutputItemDone(event, outputItems)) {
      return;
    }

    if (event.type === "response.image_generation_call.partial_image") {
      const image = partialImageEvent(event, input.responseFormat);
      if (image) {
        emit(controller, `${input.streamPrefix}.partial_image`, {
          type: `${input.streamPrefix}.partial_image`,
          ...image,
        });
      }
      return;
    }

    if (event.type === "response.completed") {
      const completedImages = extractImagesFromCompletedEvent(
        event,
        outputItems,
      );
      const payload = buildImagesStreamCompletedEvents(
        completedImages,
        input.responseFormat,
      );
      for (const item of payload) {
        emit(controller, `${input.streamPrefix}.completed`, {
          type: `${input.streamPrefix}.completed`,
          ...item,
        });
      }
      completed = true;
      input.onCompleted();
    }
  }

  return upstreamBody.pipeThrough(
    new TransformStream<Uint8Array, Uint8Array>({
      transform(chunk, controller) {
        buffer += decoder.decode(chunk, { stream: true });
        const frames = splitCompleteSseFrames(buffer);
        buffer = frames.remaining;
        try {
          for (const frame of frames.complete) {
            processFrame(frame, controller);
          }
        } catch (error) {
          fail(controller, error);
        }
      },
      flush(controller) {
        buffer += decoder.decode();
        try {
          for (const frame of parseSseFrames(buffer)) {
            processFrame(frame, controller);
          }
          if (!completed) {
            throw new Error(
              "Upstream stream ended before image generation completed",
            );
          }
        } catch (error) {
          fail(controller, error);
        }
      },
    }),
  );
}

function imageGenerationTool(
  action: ImageAction,
  model: string,
  input: Record<string, unknown>,
) {
  const tool: Record<string, unknown> = {
    type: "image_generation",
    action,
    model,
  };
  const stringFields = [
    "size",
    "quality",
    "background",
    "output_format",
    "moderation",
    ...(action === "edit" ? ["input_fidelity"] : []),
  ];
  for (const field of stringFields) {
    const value = stringValue(input[field]);
    if (value) {
      tool[field] = value;
    }
  }
  for (const field of ["output_compression", "partial_images"]) {
    const value = numberValue(input[field]);
    if (value !== null) {
      tool[field] = value;
    }
  }
  return tool;
}

function buildImagesResponsesPayload(
  prompt: string,
  images: string[],
  tool: Record<string, unknown>,
) {
  const mainModel = imagesMainModelForTool(stringValue(tool.model));
  return {
    instructions: "",
    stream: true,
    reasoning: { effort: "medium", summary: "auto" },
    parallel_tool_calls: true,
    include: ["reasoning.encrypted_content"],
    model: mainModel,
    store: false,
    tool_choice: { type: "image_generation" },
    input: [
      {
        type: "message",
        role: "user",
        content: [
          { type: "input_text", text: prompt },
          ...images
            .map((image) => image.trim())
            .filter(Boolean)
            .map((image_url) => ({ type: "input_image", image_url })),
        ],
      },
    ],
    tools: [tool],
  };
}

function extractImagesFromCompletedEvent(
  event: Record<string, unknown>,
  repairState: OutputItemRepairState = createOutputItemRepairState(),
) {
  const response = isRecord(event.response) ? event.response : event;
  const created =
    numberValue(response.created_at) ?? Math.floor(Date.now() / 1000);
  const output = repairedCompletedOutput(response, repairState);
  const results: ImageCallResult[] = [];
  for (const item of output) {
    if (!isRecord(item) || item.type !== "image_generation_call") {
      continue;
    }
    const result = stringValue(item.result);
    if (!result) {
      continue;
    }
    results.push({
      result,
      revisedPrompt: stringValue(item.revised_prompt),
      outputFormat: stringValue(item.output_format),
      size: stringValue(item.size),
      background: stringValue(item.background),
      quality: stringValue(item.quality),
    });
  }
  if (results.length === 0) {
    throw new HttpError(
      502,
      "missing_image_output",
      "Upstream did not return image output",
    );
  }
  const toolUsage = isRecord(response.tool_usage)
    ? response.tool_usage.image_gen
    : undefined;
  return {
    created,
    results,
    usage: isRecord(toolUsage) ? toolUsage : null,
    firstMeta: results[0],
  };
}

function createOutputItemRepairState(): OutputItemRepairState {
  return { byIndex: new Map(), fallback: [] };
}

function collectOutputItemDone(
  event: Record<string, unknown>,
  state: OutputItemRepairState,
) {
  if (event.type !== "response.output_item.done" || !isRecord(event.item)) {
    return false;
  }
  const outputIndex = numberValue(event.output_index);
  if (outputIndex !== null) {
    state.byIndex.set(outputIndex, event.item);
  } else {
    state.fallback.push(event.item);
  }
  return true;
}

function repairedCompletedOutput(
  response: Record<string, unknown>,
  state: OutputItemRepairState,
) {
  const output = Array.isArray(response.output) ? response.output : [];
  if (output.length > 0 || (state.byIndex.size === 0 && state.fallback.length === 0)) {
    return output;
  }
  return [
    ...[...state.byIndex.entries()]
      .sort(([left], [right]) => left - right)
      .map(([, item]) => item),
    ...state.fallback,
  ];
}

function buildImagesApiResponse(
  input: ReturnType<typeof extractImagesFromCompletedEvent>,
  responseFormat: "b64_json" | "url",
) {
  const output: Record<string, unknown> = {
    created: input.created,
    data: input.results.map((image) =>
      imageApiDataItem(image, responseFormat),
    ),
  };
  if (input.firstMeta.background) {
    output.background = input.firstMeta.background;
  }
  if (input.firstMeta.outputFormat) {
    output.output_format = input.firstMeta.outputFormat;
  }
  if (input.firstMeta.quality) {
    output.quality = input.firstMeta.quality;
  }
  if (input.firstMeta.size) {
    output.size = input.firstMeta.size;
  }
  if (input.usage) {
    output.usage = input.usage;
  }
  return output;
}

function buildImagesStreamCompletedEvents(
  input: ReturnType<typeof extractImagesFromCompletedEvent>,
  responseFormat: "b64_json" | "url",
) {
  return input.results.map((image) => {
    const item = imageApiDataItem(image, responseFormat);
    if (input.usage) {
      item.usage = input.usage;
    }
    return item;
  });
}

function imageApiDataItem(
  image: ImageCallResult,
  responseFormat: "b64_json" | "url",
) {
  const item: Record<string, unknown> = {};
  if (responseFormat === "url") {
    item.url = `data:${mimeTypeFromOutputFormat(image.outputFormat)};base64,${
      image.result
    }`;
  } else {
    item.b64_json = image.result;
  }
  if (image.revisedPrompt) {
    item.revised_prompt = image.revisedPrompt;
  }
  return item;
}

function partialImageEvent(
  event: Record<string, unknown>,
  responseFormat: "b64_json" | "url",
) {
  const b64 = stringValue(event.partial_image_b64);
  if (!b64) {
    return null;
  }
  const item: Record<string, unknown> = {
    partial_image_index: numberValue(event.partial_image_index) ?? 0,
  };
  if (responseFormat === "url") {
    item.url = `data:${mimeTypeFromOutputFormat(
      stringValue(event.output_format),
    )};base64,${b64}`;
  } else {
    item.b64_json = b64;
  }
  return item;
}

function collectJsonImages(input: Record<string, unknown>) {
  const images = Array.isArray(input.images) ? input.images : [];
  return images.map(imageUrlFromRecord).filter(Boolean);
}

function imageUrlFromRecord(input: unknown) {
  if (typeof input === "string") {
    return input.trim();
  }
  if (!isRecord(input)) {
    return "";
  }
  if (typeof input.image_url === "string") {
    return input.image_url.trim();
  }
  if (isRecord(input.image_url)) {
    return stringValue(input.image_url.url);
  }
  return stringValue(input.url);
}

function imagesMainModelForTool(toolModel: string) {
  const index = toolModel.lastIndexOf("/");
  if (index > 0 && index < toolModel.length - 1) {
    const prefix = toolModel.slice(0, index).trim();
    if (prefix) {
      return `${prefix}/${DEFAULT_IMAGES_MAIN_MODEL}`;
    }
  }
  return DEFAULT_IMAGES_MAIN_MODEL;
}

function normalizeImagesModel(value: unknown) {
  const model = stringValue(value) || DEFAULT_IMAGES_TOOL_MODEL;
  const base = model.slice(model.lastIndexOf("/") + 1).toLowerCase();
  if (base === DEFAULT_IMAGES_TOOL_MODEL) {
    return model;
  }
  throw new HttpError(
    400,
    "unsupported_image_model",
    `Model ${model} is not supported on /v1/images/generations or /v1/images/edits. Use ${DEFAULT_IMAGES_TOOL_MODEL}.`,
  );
}

function normalizeImagesResponseFormat(value: unknown): "b64_json" | "url" {
  return stringValue(value).toLowerCase() === "url" ? "url" : "b64_json";
}

function parseSseFrames(text: string) {
  return text
    .replace(/\r\n/g, "\n")
    .split(/\n\n+/)
    .map((frame) => frame.trim())
    .filter(Boolean);
}

function splitCompleteSseFrames(text: string) {
  const normalized = text.replace(/\r\n/g, "\n");
  const parts = normalized.split(/\n\n+/);
  const complete = parts.slice(0, -1).filter((frame) => frame.trim());
  const remaining = parts.at(-1) || "";
  return { complete, remaining };
}

function parseSseJson(frame: string) {
  const data = frame
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter((line) => line.startsWith("data:"))
    .map((line) => line.slice(5).trim())
    .join("\n");
  if (!data || data === "[DONE]") {
    return null;
  }
  try {
    const parsed = JSON.parse(data) as unknown;
    return isRecord(parsed) ? parsed : null;
  } catch {
    throw new HttpError(
      502,
      "invalid_image_sse_json",
      "Upstream returned invalid image SSE JSON",
    );
  }
}

function mimeTypeFromOutputFormat(outputFormat: string) {
  const value = outputFormat.trim().toLowerCase();
  if (value.includes("/")) {
    return value;
  }
  switch (value) {
    case "jpg":
    case "jpeg":
      return "image/jpeg";
    case "webp":
      return "image/webp";
    case "png":
    default:
      return "image/png";
  }
}

async function fileToDataUrl(file: File) {
  if (file.size > MAX_IMAGE_EDIT_FILE_BYTES) {
    throw new HttpError(
      413,
      "image_file_too_large",
      "Image file is too large",
    );
  }
  const bytes = Buffer.from(await file.arrayBuffer());
  const mediaType = file.type || "application/octet-stream";
  return `data:${mediaType};base64,${bytes.toString("base64")}`;
}

function formRecordFromData(formData: FormData) {
  const record: Record<string, unknown> = {};
  for (const [key, value] of formData.entries()) {
    if (typeof value === "string") {
      record[key] = value;
    }
  }
  return record;
}

function isFileLike(value: unknown): value is File {
  return Boolean(
    value &&
      typeof value === "object" &&
      "arrayBuffer" in value &&
      typeof (value as File).arrayBuffer === "function",
  );
}

function booleanFormValue(value: FormDataEntryValue | null) {
  const raw = stringValue(value).toLowerCase();
  return ["1", "true", "yes", "on"].includes(raw);
}

function numberValue(value: unknown) {
  if (typeof value === "number" && Number.isFinite(value)) {
    return Math.floor(value);
  }
  if (typeof value === "string" && value.trim()) {
    const parsed = Number(value);
    return Number.isFinite(parsed) ? Math.floor(parsed) : null;
  }
  return null;
}

function stringValue(value: unknown) {
  return typeof value === "string"
    ? value.trim()
    : typeof value === "number"
      ? String(value)
      : "";
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value && typeof value === "object" && !Array.isArray(value));
}

File: server/src/upstream/codex/images.rs
use std::{collections::BTreeMap, io};

use async_stream::stream;
use axum::body::Bytes;
use chrono::Utc;
use futures_util::{Stream, StreamExt};
use serde_json::{json, Map, Value};

use crate::{
    error::{AppError, AppResult},
    upstream::codex::sse,
};

#[derive(Clone, Copy)]
pub enum ResponseFormat {
    B64Json,
    Url,
}

pub fn response_format(input: &Value) -> ResponseFormat {
    if input
        .get("response_format")
        .and_then(Value::as_str)
        .is_some_and(|value| value.eq_ignore_ascii_case("url"))
    {
        ResponseFormat::Url
    } else {
        ResponseFormat::B64Json
    }
}

pub fn generation_to_responses(input: Value) -> AppResult<Value> {
    let object = input.as_object().ok_or_else(|| {
        AppError::bad_request("invalid_json", "Request body must be a JSON object")
    })?;
    let prompt = object
        .get("prompt")
        .and_then(Value::as_str)
        .filter(|value| !value.trim().is_empty())
        .ok_or_else(|| AppError::bad_request("missing_image_prompt", "prompt is required"))?;
    Ok(build_payload("generate", prompt, Vec::new(), object))
}

pub fn edit_to_responses(input: Value) -> AppResult<Value> {
    let object = input.as_object().ok_or_else(|| {
        AppError::bad_request("invalid_json", "Request body must be a JSON object")
    })?;
    let prompt = object
        .get("prompt")
        .and_then(Value::as_str)
        .filter(|value| !value.trim().is_empty())
        .ok_or_else(|| AppError::bad_request("missing_image_prompt", "prompt is required"))?;
    let mut images = Vec::new();
    if let Some(values) = object.get("images").and_then(Value::as_array) {
        for value in values {
            if let Some(url) = image_url(value) {
                images.push(url.to_string());
            }
        }
    }
    if images.is_empty() {
        return Err(AppError::bad_request(
            "missing_image_input",
            "images[].image_url is required",
        ));
    }
    Ok(build_payload("edit", prompt, images, object))
}

pub fn response_text_to_images(text: &str, response_format: ResponseFormat) -> AppResult<Value> {
    let raw = sse::parse_codex_response_text(text).ok_or_else(|| {
        AppError::bad_gateway(
            "image_generation_incomplete",
            "Upstream stream ended before image generation completed",
        )
    })?;
    response_to_images(&raw, response_format)
}

pub fn response_to_images(raw: &Value, response_format: ResponseFormat) -> AppResult<Value> {
    let root = raw
        .get("response")
        .and_then(Value::as_object)
        .or_else(|| raw.as_object())
        .ok_or_else(|| {
            AppError::bad_gateway("invalid_image_response", "Invalid Codex image response")
        })?;
    let results = image_results(root)?;
    let first = results.first();
    let mut output = json!({
        "created": number_field(root, "created_at").unwrap_or_else(|| Utc::now().timestamp()),
        "data": results
            .iter()
            .map(|result| image_data_item(result, response_format))
            .collect::<Vec<_>>()
    });
    if let Some(value) = first
        .map(|value| value.background.as_str())
        .filter(|value| !value.is_empty())
    {
        output["background"] = Value::String(value.to_string());
    }
    if let Some(value) = first
        .map(|value| value.output_format.as_str())
        .filter(|value| !value.is_empty())
    {
        output["output_format"] = Value::String(value.to_string());
    }
    if let Some(value) = first
        .map(|value| value.quality.as_str())
        .filter(|value| !value.is_empty())
    {
        output["quality"] = Value::String(value.to_string());
    }
    if let Some(value) = first
        .map(|value| value.size.as_str())
        .filter(|value| !value.is_empty())
    {
        output["size"] = Value::String(value.to_string());
    }
    if let Some(usage) = root
        .get("tool_usage")
        .and_then(|value| value.get("image_gen"))
        .cloned()
    {
        output["usage"] = usage;
    }
    Ok(output)
}

pub fn images_sse_stream<S>(
    upstream: S,
    response_format: ResponseFormat,
    stream_prefix: &'static str,
) -> impl Stream<Item = Result<Bytes, io::Error>>
where
    S: Stream<Item = Result<Bytes, io::Error>> + Send + 'static,
{
    stream! {
        let mut upstream = Box::pin(upstream);
        let mut buffer = String::new();
        let mut state = ImagesSseState::new(response_format, stream_prefix);
        while let Some(chunk) = upstream.next().await {
            let chunk = match chunk {
                Ok(chunk) => chunk,
                Err(error) => {
                    yield Err(error);
                    return;
                }
            };
            let text = String::from_utf8_lossy(&chunk);
            for event in sse::push_json_events(&mut buffer, &text) {
                for frame in state.handle_event(event) {
                    yield Ok(Bytes::from(frame));
                }
            }
        }
        for event in sse::flush_json_events(&mut buffer) {
            for frame in state.handle_event(event) {
                yield Ok(Bytes::from(frame));
            }
        }
        if !state.done {
            yield Ok(Bytes::from(image_error_frame("Upstream stream ended before image generation completed")));
        }
    }
}

#[derive(Default)]
struct ImageResult {
    result: String,
    revised_prompt: String,
    output_format: String,
    size: String,
    background: String,
    quality: String,
}

struct ImagesSseState {
    response_format: ResponseFormat,
    stream_prefix: &'static str,
    output_by_index: BTreeMap<i64, Value>,
    output_fallback: Vec<Value>,
    done: bool,
}

impl ImagesSseState {
    fn new(response_format: ResponseFormat, stream_prefix: &'static str) -> Self {
        Self {
            response_format,
            stream_prefix,
            output_by_index: BTreeMap::new(),
            output_fallback: Vec::new(),
            done: false,
        }
    }

    fn handle_event(&mut self, event: Value) -> Vec<String> {
        match event
            .get("type")
            .and_then(Value::as_str)
            .unwrap_or_default()
        {
            "response.output_item.done" => {
                if let Some(item) = event.get("item").cloned() {
                    if let Some(index) = event.get("output_index").and_then(Value::as_i64) {
                        self.output_by_index.insert(index, item);
                    } else {
                        self.output_fallback.push(item);
                    }
                }
                Vec::new()
            }
            "response.image_generation_call.partial_image" => self.partial_image_frame(&event),
            "response.completed" => self.completed_frames(event),
            _ => Vec::new(),
        }
    }

    fn partial_image_frame(&self, event: &Value) -> Vec<String> {
        let Some(b64) = event.get("partial_image_b64").and_then(Value::as_str) else {
            return Vec::new();
        };
        let mut payload = json!({
            "type": format!("{}.partial_image", self.stream_prefix),
            "partial_image_index": event.get("partial_image_index").and_then(Value::as_i64).unwrap_or(0),
        });
        set_image_data(
            &mut payload,
            b64,
            event
                .get("output_format")
                .and_then(Value::as_str)
                .unwrap_or("png"),
            self.response_format,
        );
        vec![event_frame(
            &format!("{}.partial_image", self.stream_prefix),
            payload,
        )]
    }

    fn completed_frames(&mut self, event: Value) -> Vec<String> {
        let mut response = event.get("response").cloned().unwrap_or(event);
        let output_missing = response
            .get("output")
            .and_then(Value::as_array)
            .map(Vec::is_empty)
            .unwrap_or(true);
        if output_missing {
            let output = self
                .output_by_index
                .values()
                .cloned()
                .chain(self.output_fallback.iter().cloned())
                .collect::<Vec<_>>();
            if !output.is_empty() {
                if let Some(object) = response.as_object_mut() {
                    object.insert("output".to_string(), Value::Array(output));
                }
            }
        }
        let response = match response_to_images(&response, self.response_format) {
            Ok(response) => response,
            Err(error) => return vec![image_error_frame(&error.to_string())],
        };
        self.done = true;
        response
            .get("data")
            .and_then(Value::as_array)
            .into_iter()
            .flatten()
            .map(|item| {
                let mut payload = item.clone();
                if let Some(object) = payload.as_object_mut() {
                    object.insert(
                        "type".to_string(),
                        Value::String(format!("{}.completed", self.stream_prefix)),
                    );
                    if let Some(usage) = response.get("usage") {
                        object.insert("usage".to_string(), usage.clone());
                    }
                }
                event_frame(&format!("{}.completed", self.stream_prefix), payload)
            })
            .collect()
    }
}

fn build_payload(
    action: &str,
    prompt: &str,
    images: Vec<String>,
    object: &Map<String, Value>,
) -> Value {
    let stream = object
        .get("stream")
        .and_then(Value::as_bool)
        .unwrap_or(false);
    let image_model = object
        .get("model")
        .and_then(Value::as_str)
        .unwrap_or("gpt-image-2");
    let mut content = vec![json!({ "type": "input_text", "text": prompt })];
    for image in images {
        content.push(json!({ "type": "input_image", "image_url": image }));
    }
    let mut tool = json!({ "type": "image_generation", "action": action, "model": image_model, "output_format": "png" });
    for key in [
        "size",
        "quality",
        "background",
        "moderation",
        "output_compression",
        "partial_images",
        "input_fidelity",
    ] {
        if let Some(value) = object.get(key) {
            tool[key] = value.clone();
        }
    }
    if let Some(mask) = object.get("mask").and_then(image_url) {
        tool["input_image_mask"] = json!({ "image_url": mask });
    }
    json!({
        "model": "gpt-5.4-mini",
        "stream": stream,
        "instructions": "",
        "store": false,
        "parallel_tool_calls": true,
        "reasoning": { "effort": "medium", "summary": "auto" },
        "input": [{ "type": "message", "role": "user", "content": content }],
        "tools": [tool],
        "tool_choice": { "type": "image_generation" },
        "include": ["reasoning.encrypted_content"]
    })
}

fn image_results(root: &Map<String, Value>) -> AppResult<Vec<ImageResult>> {
    let mut results = Vec::new();
    for item in root
        .get("output")
        .and_then(Value::as_array)
        .into_iter()
        .flatten()
    {
        let Some(object) = item.as_object() else {
            continue;
        };
        if object.get("type").and_then(Value::as_str) != Some("image_generation_call") {
            continue;
        }
        let Some(result) = object.get("result").and_then(Value::as_str) else {
            continue;
        };
        results.push(ImageResult {
            result: result.to_string(),
            revised_prompt: string_field(object, "revised_prompt").unwrap_or_default(),
            output_format: string_field(object, "output_format")
                .unwrap_or_else(|| "png".to_string()),
            size: string_field(object, "size").unwrap_or_default(),
            background: string_field(object, "background").unwrap_or_default(),
            quality: string_field(object, "quality").unwrap_or_default(),
        });
    }
    if results.is_empty() {
        return Err(AppError::bad_gateway(
            "missing_image_output",
            "Upstream did not return image output",
        ));
    }
    Ok(results)
}

fn image_data_item(result: &ImageResult, response_format: ResponseFormat) -> Value {
    let mut item = json!({});
    set_image_data(
        &mut item,
        &result.result,
        &result.output_format,
        response_format,
    );
    if !result.revised_prompt.is_empty() {
        item["revised_prompt"] = Value::String(result.revised_prompt.clone());
    }
    item
}

fn set_image_data(
    item: &mut Value,
    b64: &str,
    output_format: &str,
    response_format: ResponseFormat,
) {
    match response_format {
        ResponseFormat::B64Json => item["b64_json"] = Value::String(b64.to_string()),
        ResponseFormat::Url => {
            item["url"] = Value::String(format!(
                "data:{};base64,{}",
                mime_type_from_output_format(output_format),
                b64
            ));
        }
    }
}

fn image_url(value: &Value) -> Option<&str> {
    if let Some(value) = value.as_str() {
        return Some(value.trim()).filter(|value| !value.is_empty());
    }
    value
        .get("image_url")
        .and_then(|value| value.get("url").or(Some(value)))
        .or_else(|| value.get("url"))
        .and_then(Value::as_str)
        .map(str::trim)
        .filter(|value| !value.is_empty())
}

fn event_frame(event: &str, data: Value) -> String {
    format!("event: {event}\ndata: {data}\n\n")
}

fn image_error_frame(message: &str) -> String {
    event_frame(
        "error",
        json!({ "error": { "message": message, "type": "stream_error", "code": "upstream_image_stream_error" } }),
    )
}

fn mime_type_from_output_format(output_format: &str) -> &'static str {
    match output_format.trim().to_ascii_lowercase().as_str() {
        "jpg" | "jpeg" | "image/jpeg" => "image/jpeg",
        "webp" | "image/webp" => "image/webp",
        _ => "image/png",
    }
}

fn string_field(object: &Map<String, Value>, key: &str) -> Option<String> {
    object
        .get(key)
        .and_then(Value::as_str)
        .map(str::to_string)
        .filter(|value| !value.trim().is_empty())
}

fn number_field(object: &Map<String, Value>, key: &str) -> Option<i64> {
    object.get(key).and_then(|value| match value {
        Value::Number(value) => value
            .as_i64()
            .or_else(|| value.as_u64().map(|value| value as i64)),
        Value::String(value) => value.parse::<i64>().ok(),
        _ => None,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn converts_codex_image_response_to_openai_images_shape() {
        let raw = json!({
            "id": "resp_img",
            "object": "response",
            "created_at": 123,
            "status": "completed",
            "output": [{
                "type": "image_generation_call",
                "result": "ZmFrZQ==",
                "revised_prompt": "cat",
                "output_format": "png",
                "size": "1024x1024",
                "quality": "auto"
            }],
            "tool_usage": { "image_gen": { "images": 1 } }
        });

        let out = response_to_images(&raw, ResponseFormat::B64Json).unwrap();

        assert_eq!(out["data"][0]["b64_json"], "ZmFrZQ==");
        assert_eq!(out["data"][0]["revised_prompt"], "cat");
        assert_eq!(out["size"], "1024x1024");
        assert_eq!(out["usage"]["images"], 1);
    }
}
